flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -0,0 +1,609 @@
+			}
+		}
+		return numUnfulfilled;
+	}
+
+	/**
+	 * Allocates the given slot. This entails sending a registration message to the task manager and treating failures.
+	 *
+	 * @param taskManagerSlot slot to allocate
+	 * @param jobId job for which the slot should be allocated for
+	 * @param targetAddress address of the job master
+	 * @param resourceProfile resource profile for the requirement for which the slot is used
+	 */
+	private void allocateSlot(TaskManagerSlotInformation taskManagerSlot, JobID jobId, String targetAddress, ResourceProfile resourceProfile) {
+		TaskExecutorConnection taskExecutorConnection = taskManagerSlot.getTaskManagerConnection();
+		TaskExecutorGateway gateway = taskExecutorConnection.getTaskExecutorGateway();
+
+		final CompletableFuture<Acknowledge> completableFuture = new CompletableFuture<>();
+		final SlotID slotId = taskManagerSlot.getSlotId();
+		final InstanceID instanceId = taskManagerSlot.getInstanceId();
+
+		LOG.debug("Starting allocation of slot {} for job {} with resource profile {}.", taskManagerSlot.getSlotId(), jobId, resourceProfile);
+
+		slotTracker.notifyAllocationStart(slotId, jobId);
+
+		if (!taskExecutorManager.isTaskManagerRegistered(instanceId)) {
+			throw new IllegalStateException("Could not find a registered task manager for instance id " +
+				instanceId + '.');
+		}
+
+		taskExecutorManager.markUsed(instanceId);
+
+		// RPC call to the task manager
+		CompletableFuture<Acknowledge> requestFuture = gateway.requestSlot(
+			slotId,
+			jobId,
+			new AllocationID(),
+			resourceProfile,
+			targetAddress,
+			resourceManagerId,
+			taskManagerRequestTimeout);
+
+		requestFuture.whenComplete(
+			(Acknowledge acknowledge, Throwable throwable) -> {
+				if (acknowledge != null) {
+					completableFuture.complete(acknowledge);
+				} else {
+					completableFuture.completeExceptionally(throwable);
+				}
+			});
+
+		completableFuture.whenCompleteAsync(
+			(Acknowledge acknowledge, Throwable throwable) -> {
+				try {
+					if (acknowledge != null) {
+						LOG.trace("Completed allocation of slot {} for job {}.", slotId, jobId);
+						slotTracker.notifyAllocationComplete(slotId, jobId);
+					} else {
+						if (throwable instanceof SlotOccupiedException) {
+							SlotOccupiedException exception = (SlotOccupiedException) throwable;
+							LOG.debug("Tried allocating slot {} for job {}, but it was already allocated for job {}.", slotId, jobId, exception.getJobId());
+							// report as a slot status to force the state transition
+							// this could be a problem if we ever assume that the task executor always reports about all slots
+							slotTracker.notifySlotStatus(Collections.singleton(new SlotStatus(slotId, taskManagerSlot.getResourceProfile(), exception.getJobId(), exception.getAllocationId())));
+						} else {
+							if (throwable instanceof CancellationException) {
+								LOG.debug("Cancelled allocation of slot {} for job {}.", slotId, jobId, throwable);
```

Comment:
Where are we cancelling `completableFuture`?

Reply:
Whenever a pending slot switches to another state (see `createSlotStatusUpdateListener`)

Reply:
I think at the moment, we don't register these futures and, hence, it should never happen.